### ISSUE TITLE
[Question] Rate limits, internals and async

### ISSUE THREAD

**Issue by ryoko7** [exislow/tidal-dl-ng#244]
**Note**, these questions are related to programmatic usage of tidal-dl-ng. _I do not need help_, but any answers would be appreciated.
1. Does tidal-dl-ng apply rate limits to downloading? If so, is it possible that f28d07b651a0d69229bfb6f1a1bcd3525ed99933 will eventually use download speed as a way to rate limit?

2. Is it safe to assume that gui.py and cli.py control the entire process? If so, is it suggested to create a different file specifically for programmatic usage? In that case, I may make a pull request after making that file.

3. Finally, would tidal-dl-ng work with async code?

### What operating system are you seeing the problem on?
None

### Your settings
N/A. Help not needed.

**Comment by exislow** [comment]
May I ask what do you actually plan to do? I do not quite understand what your goal is.

**Comment by ryoko7** [comment]
My plan is to use tidal-dl-ng with a rate limit. However, it's a bit tough to implement outside of the actual downloader.
I was hoping to use async to regularly check my battery level, but I'll have to settle with checking it in between Download function calls. Regarding point # 2, I'm going to create a programmatic interface to do essentially the same things that cli.py/gui.py does, for my mini-project. I think I will have to integrate directly with tidalapi to somehow automatically locate artists/songs, though. 

TL;DR I'm working on a script to get all of my Spotify songs from Tidal. 
I don't want to overwhelm Tidal due to the large volume and it would take far too long to locate each artist manually (3000+ artists)

**Comment by exislow** [comment]
Since this project is based on `tidalapi`, maybe it would be best, to use `tidalapi` directly in your case.

Also there are services like this: https://www.tunemymusic.com/de/transfer/spotify-to-tidal

Maybe this helps: https://github.com/exislow/tidal-dl-ng/blob/8b895ce1846762fad7789d6fee3392211ca69c76/tidal_dl_ng/model/cfg.py#L41-L42

Also don't be mad, but I do not have enough free time to describe you this code base line by line. I mean, you have access to whole `tidal-dl-ng` code. Just fire up your debugger and see what happens, if you are to lazy to read the whole code yourself :-)

**Comment by ryoko7** [comment]
That's no problem. Thank you for your assistance nonetheless and the work you do. In general, I trust tidal-dl-ng will work perfect :-). I may need to search the artists directly using tidalapi, but I think your codebase will work great for my use case. Again, thank you
